⬆️ update MacOS runner versions in GitHub CI pipeline

kevinmatthes opened the pull request:
@ttytm

This PR fixes #348.  I replaced ``macos-11`` with the newer ``macos-12`` and ``macos-13`` settings.  Furthermore, in ``release.yml``, I also added further compilation targets such that MacOS versions 12 and 13 will both be targeted with ``x86_64`` as well as ``arm64`` architectures.

I would like to ask you to review these changes.

kevinmatthes commented:
Drafted due to some bugs to be fixed.

kevinmatthes commented:
@ttytm

I am sorry, my changes keep failing.  Do you have an idea what could have gone wrong?

ttytm commented:
Hey @kevinmatthes,

Thanks for your efforts!

Yes, the macos-11 runner images were removed after a period of downtime.

Apparently it's still a problem using the toolchain out of the box on the newer mac intel runners.
The same failures were the reason I fall back to the old runner for x86 compilation instead of using the macos-13 runner already.

Manually setting up the rust toolchain might work.

ttytm commented on .github/workflows/ci.yml:
```diff
@@ -41,7 +41,7 @@ jobs:
       fail-fast: false
       matrix:
-        os: [macos-11, macos-latest, ubuntu-latest, windows-latest]
+        os: [macos-12, macos-13, macos-latest, ubuntu-latest, windows-latest]
```
One x86 runner will suffice. If we get the 13 working, let's use it.

macos-12 reaches end of life before the end of the year and the binary compiled by the more modern runner should still be compatible.

kevinmatthes replied:
🔧 ✅

kevinmatthes commented:
@ttytm

Thank you for your review! 👍 

I have already tried to setup Rust manually in 08a531fe204d22ac3c55dd1c82f669c98edbb82b where I called the installation script for ``rustup``.  It did not work as expected, so I switched to the current solution using a GitHub Action.

I will keep thinking about this issue, maybe I can find some other approach...

kevinmatthes commented:
This comment is just for documentation of the last commits.

- I removed MacOS 12 runners in both release and CI workflows.
- I switched back to manual Rust re-installation on the MacOS 13 hosted runner as this might be a better starting point for a solution to the problem.
- I added an explicit directory creation instruction to ensure that the target directory actually exists.

The ``rustup`` approach always fails when it comes to copying the downloaded Rust components to ``~/.cargo/bin/rustup``.  As the directory is now guaranteed to exist, there must be some issue about the source directory...

ttytm commented:
Thanks again!

kevinmatthes commented:
Adding super user permissions to the execution of the ``rustup`` installation script does not help, either...

ttytm commented on .github/workflows/ci.yml:
```diff
@@ -56,6 +56,11 @@ jobs:
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
```
```suggestion
          key: ${{ matrix.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
```

This ones the culprit. Using the same cache key for the arm and intel runner. Setting a dedicated key per runner makes it work as expected.

Sample run (only ci.yml updated):
https://github.com/ttytm/wthrr-the-weathercrab/actions/runs/10267829477/job/28409372819

kevinmatthes replied:
It finally works!  Thank you very much for your help, @ttytm! 🎉

It seems as if we would not need to re-install Rust, either.  Feel free to remove that step, if you like.
